File: gateway/src/DijiPeople.Gateway.Host/Connectors/ZkTeco/ZkTecoLegacyConnector.cs
using System.Globalization;
using System.Text.RegularExpressions;

using DijiPeople.Gateway.Cloud;
using DijiPeople.Gateway.Storage;

using Microsoft.Extensions.Logging;

namespace DijiPeople.Gateway.Connectors.ZkTeco;

/// <summary>
/// ZKTeco standalone terminals over the legacy zkemkeeper SDK.
///
/// Every safety property proved by the diagnostic POC is preserved here, because
/// this adapter runs the same worker binary rather than reimplementing the COM
/// calls:
///
///   x86 enforced three ways · read-only method allowlist · Disconnect and COM
///   release in a finally · no biometric read · device password discarded inside
///   the worker · no log clearing · no clock mutation · no user mutation during
///   attendance sync.
///
/// ReadLastestLogData, ReadMark, SetLastCount, ClearGLog and ClearData are not
/// on the worker's allowlist and are never invoked. That matters beyond
/// tidiness: the customer's V2011 software reads the same terminal and shares
/// any device-side read marker, so advancing one would make V2011's next
/// incremental download silently skip records with no way to undo it.
///
/// WHAT THIS ADAPTER CANNOT DO. The installed SDK exposes neither a time-bounded
/// read nor a new-only read, confirmed against the physical device. Every poll
/// therefore returns the terminal's entire stored history. Deduplication and the
/// import window both live on the DijiPeople side, which is why the local
/// fingerprint store is not an optimisation but the mechanism.
/// </summary>
internal sealed class ZkTecoLegacyConnector : IGatewayAttendanceConnector
{
    private readonly ZkTecoWorkerClient _worker;
    private readonly ILogger<ZkTecoLegacyConnector> _logger;

    /// <summary>Matches AttendanceIntegration.connectorType in DijiPeople.</summary>
    public string ConnectorType => "zkteco-legacy-tcp";

    /// <summary>
    /// WRITE_USERS is deliberately ABSENT.
    ///
    /// The SDK exposes a user write path and the API's connector definition
    /// declares it as an experimental capability, but it has never been executed
    /// against a physical terminal. Until that validation happens on the
    /// customer's device, this adapter must not advertise a capability the
    /// scheduler would then act on unattended. Adding it here before the
    /// hardware test would make the gate a formality.
    /// </summary>
    public IReadOnlySet<string> Capabilities { get; } = new HashSet<string>(StringComparer.Ordinal)
    {
        "READ_DEVICE_INFO",
        "READ_USERS",
        "READ_ATTENDANCE",
        "POLL_EVENTS",
        "DEVICE_TIME",
        "PUNCH_STATE",
        "WORK_CODE",
        "LOCAL_GATEWAY_REQUIRED",
    };

    /// <summary>Device wall clock. Anything else is not a timestamp we will send.</summary>
    private static readonly Regex LocalTimestamp = new(
        @"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}$",
        RegexOptions.Compiled);

    public ZkTecoLegacyConnector(
        ZkTecoWorkerClient worker,
        ILogger<ZkTecoLegacyConnector> logger)
    {
        _worker = worker;
        _logger = logger;
    }

    // ------------------------------------------------------------- verify

    public async Task<DeviceVerificationResult> VerifyDeviceAsync(
        ConnectorDeviceContext context,
        CancellationToken cancellationToken)
    {
        // --device-info connects, reads identity and the clock, and disconnects.
        // It writes nothing and touches no log marker.
        var invocation = await _worker.InvokeAsync(
            BuildArguments(context, "--device-info"),
            cancellationToken);

        if (!invocation.Succeeded || invocation.Result is null)
        {
            return new DeviceVerificationResult(
                Connected: false,
                LatencyMs: invocation.DurationMs,
                SerialNumber: null,
                Model: null,
                FirmwareVersion: null,
                Platform: null,
                MacAddress: null,
                DeviceTimeLocal: null,
                ErrorCode: invocation.ErrorCode ?? "DEVICE_UNREACHABLE",
                ErrorMessage: invocation.ErrorMessage);
        }

        var result = invocation.Result;
        var device = result.Device;

        return new DeviceVerificationResult(
            Connected: result.Connection?.Connected ?? false,
            // The SDK's own connect timing, not the whole process lifetime —
            // process start-up would otherwise show as device latency.
            LatencyMs: (int)(result.Connection?.ConnectDurationMs ?? invocation.DurationMs),
            SerialNumber: Clean(device?.SerialNumber),
            Model: Clean(device?.Model),
            FirmwareVersion: Clean(device?.FirmwareVersion),
            Platform: Clean(device?.Platform),
            MacAddress: Clean(device?.MacAddress),
            DeviceTimeLocal: ValidLocalTimestamp(device?.DeviceTimeLocal),
            ErrorCode: null,
            ErrorMessage: null);
    }

    // ------------------------------------------------------------ discover

    public async Task<UserDiscoveryResult> DiscoverUsersAsync(
        ConnectorDeviceContext context,
        CancellationToken cancellationToken)
    {
        var arguments = BuildArguments(context, "--users");
        arguments.Add("--max-users");
        arguments.Add(context.MaxRecords.ToString(CultureInfo.InvariantCulture));

        var invocation = await _worker.InvokeAsync(arguments, cancellationToken);

        if (!invocation.Succeeded || invocation.Result is null)
        {
            return new UserDiscoveryResult(
                false,
                Array.Empty<DiscoveredUser>(),
                null,
                invocation.ErrorCode ?? "READ_USERS_FAILED",
                invocation.ErrorMessage);
        }

        var users = new List<DiscoveredUser>();

        foreach (var user in invocation.Result.Users ?? new List<WorkerUser>())
        {
            var externalUserId = Clean(user.ExternalUserId);
            if (externalUserId is null)
            {
                // A directory slot with no identifier is not a person we can map.
                continue;
            }

            // Explicit projection, not a copy of the worker object. Only these
            // four fields exist on the wire, so nothing else can travel even if
            // the worker contract grows a field later.
            users.Add(new DiscoveredUser(
                externalUserId,
                Clean(user.Name),
                user.PrivilegeRaw,
                user.Enabled));
        }

        return new UserDiscoveryResult(
            true,
            users,
            Clean(invocation.Result.Device?.SerialNumber),
            null,
            null);
    }

    // ---------------------------------------------------------- attendance

    public async Task<AttendanceReadResult> ReadAttendanceAsync(
        ConnectorDeviceContext context,
        CancellationToken cancellationToken)
    {
        var arguments = BuildArguments(context, "--attendance");
        // A safety ceiling on the enumeration loop, not an incremental read:
        // ReadGeneralLogData has already buffered the device's whole history by
        // the time this bounds anything. Treating it as "only fetch N" would be
        // a misreading with real consequences.
        arguments.Add("--max-attendance");
        arguments.Add(context.MaxRecords.ToString(CultureInfo.InvariantCulture));

        var invocation = await _worker.InvokeAsync(arguments, cancellationToken);

        if (!invocation.Succeeded || invocation.Result is null)
        {
            return new AttendanceReadResult(
                false,
                Array.Empty<ObservedPunch>(),
                null,
                null,
                invocation.ErrorCode ?? "READ_ATTENDANCE_FAILED",
                invocation.ErrorMessage);
        }

        var serial = Clean(invocation.Result.Device?.SerialNumber);
        var punches = new List<ObservedPunch>();
        var malformed = 0;

        foreach (var punch in invocation.Result.Attendance ?? new List<WorkerPunch>())
        {
            var externalUserId = Clean(punch.ExternalUserId);
            var occurredAtLocal = ValidLocalTimestamp(punch.OccurredAtLocal);

            if (externalUserId is null || occurredAtLocal is null)
            {
                malformed++;
                continue;
            }

            punches.Add(new ObservedPunch(
                externalUserId,
                occurredAtLocal,
                punch.VerificationModeRaw,
                punch.PunchStateRaw,
                punch.WorkCodeRaw,
                EventFingerprint.Compute(
                    serial,
                    externalUserId,
                    occurredAtLocal,
                    punch.VerificationModeRaw,
                    punch.PunchStateRaw,
                    punch.WorkCodeRaw)));
        }

        if (malformed > 0)
        {
            // Dropped rather than guessed at. A punch with no user or no usable
            // timestamp cannot be attributed to anyone or placed in time, and
            // inventing either would put a fiction into payroll evidence.
            _logger.LogWarning(
                "Device {Device} returned {Count} attendance record(s) with no usable user or timestamp; they were not queued.",
                context.DeviceName,
                malformed);
        }

        return new AttendanceReadResult(
            true,
            punches,
            serial,
            ValidLocalTimestamp(invocation.Result.Device?.DeviceTimeLocal),
            null,
            null);
    }

    // --------------------------------------------------------- provisioning

    /// <summary>
    /// Refuses, by design, until the write path is validated on hardware.
    ///
    /// This is not a stub standing in for missing code — the transport, the job
    /// lease, the claim and the reporting path are all built and exercised. What
    /// is missing is a physical test on the customer's terminal proving which
    /// SDK write calls are safe and what they do to a device the customer's
    /// V2011 software also manages. Guessing at that against a production
    /// terminal is the one thing this phase must not do, so the refusal is the
    /// correct behaviour and is reported honestly rather than silently
    /// succeeding.
    /// </summary>
    public Task<ProvisioningExecutionResult> ProvisionUserAsync(
        ConnectorDeviceContext context,
        ProvisioningJobPayload payload,
        string operation,
        CancellationToken cancellationToken)
    {
        _logger.LogWarning(
            "Refused provisioning job ({Operation}) for device {Device}: ZKTeco write-back is not certified.",
            operation,
            context.DeviceName);

        return Task.FromResult(new ProvisioningExecutionResult(
            Succeeded: false,
            ResultExternalUserId: null,
            ErrorCode: "WRITE_NOT_CERTIFIED",
            ErrorMessage:
            "Writing users to ZKTeco terminals has not been validated against physical hardware, so this gateway will not attempt it."));
    }

    // ---------------------------------------------------------------- helpers

    /// <summary>
    /// Builds the worker command line for one device.
    ///
    /// The comm key comes from the integration's encrypted configuration and is
    /// omitted entirely when it is zero — the reference K50's setting — which
    /// keeps the exact call sequence the POC validated. It is masked whenever
    /// this list is logged.
    /// </summary>
    private static List<string> BuildArguments(ConnectorDeviceContext context, string mode)
    {
        var arguments = new List<string>
        {
            mode,
            "--host", context.Host,
            "--port", context.Port.ToString(CultureInfo.InvariantCulture),
            "--machine-number", context.MachineNumber.ToString(CultureInfo.InvariantCulture),
        };

        var commKey = ReadCommKey(context.Configuration);
        if (commKey > 0)
        {
            arguments.Add("--comm-key");
            arguments.Add(commKey.ToString(CultureInfo.InvariantCulture));
        }

        return arguments;
    }

    /// <summary>
    /// Reads the comm key from connector configuration.
    ///
    /// Tolerant of both a JSON number and a string, because the field is a
    /// secret and secrets round-trip through encrypted JSON where a number can
    /// arrive quoted. An unparseable value becomes 0, which means "no key" —
    /// the same as not configuring one, and safely wrong rather than sending
    /// garbage to the terminal.
    /// </summary>
    internal static int ReadCommKey(IReadOnlyDictionary<string, object?> configuration)
    {
        if (!configuration.TryGetValue("commKey", out var raw) || raw is null)
        {
            return 0;
        }

        return raw switch
        {
            int value => value,
            long value => (int)value,
            double value => (int)value,
            string text when int.TryParse(
                text.Trim(),
                NumberStyles.Integer,
                CultureInfo.InvariantCulture,
                out var parsed) => parsed,
            System.Text.Json.JsonElement element => ReadCommKeyFromJson(element),
            _ => 0,
        };
    }

    private static int ReadCommKeyFromJson(System.Text.Json.JsonElement element) =>
        element.ValueKind switch
        {
            System.Text.Json.JsonValueKind.Number when element.TryGetInt32(out var number) => number,
            System.Text.Json.JsonValueKind.String when int.TryParse(
                element.GetString(),
                NumberStyles.Integer,
                CultureInfo.InvariantCulture,
                out var parsed) => parsed,
            _ => 0,
        };

    private static string? Clean(string? value)
    {
        if (string.IsNullOrWhiteSpace(value)) return null;
        var trimmed = value.Trim();
        return trimmed.Length > 200 ? trimmed[..200] : trimmed;
    }

    /// <summary>
    /// Accepts a device timestamp only in the exact wall-clock form.
    ///
    /// No parsing into a DateTime, no normalisation, no offset. The terminal
    /// states no timezone, so treating its string as a UTC instant — which
    /// DateTime.Parse would happily do for some inputs — would shift every punch
    /// by the gateway machine's offset.
    /// </summary>
    private static string? ValidLocalTimestamp(string? value)
    {
        if (string.IsNullOrWhiteSpace(value)) return null;
        var trimmed = value.Trim();
        return LocalTimestamp.IsMatch(trimmed) ? trimmed : null;
    }
}

File: services/api/src/modules/attendance-integrations/connectors/zkteco-legacy.connector.ts
import {
  AttendanceConnectionMode,
  AttendanceProvider,
  AttendanceSyncIntervalUnit,
} from '@prisma/client';

import type { AttendanceConnectorDefinition } from './connector.types';

/**
 * ZKTeco legacy standalone terminals reached over TCP through the `zkemkeeper`
 * COM SDK, running inside an on-premise gateway.
 *
 * Every capability and limit below was established against a physical ZKTeco K50
 * (serial A2QO221160250, firmware 8.0.4.2-20200723) by the diagnostic worker in
 * `tools/zkteco-poc`, not from vendor documentation:
 *
 *   - Connect_Net, GetSerialNumber, ReadAllUserID + SSR_GetAllUserInfo,
 *     ReadGeneralLogData + SSR_GetGeneralLogData and Disconnect all succeeded.
 *   - 59 device users and raw attendance transactions were retrieved.
 *   - The installed component exposes 241 methods. `ReadTimeGLogData`
 *     (time-bounded) and `ReadNewGLogData` (new-only) are BOTH ABSENT, so this
 *     device family offers no incremental read: every poll returns the entire
 *     history, observed back to October 2022.
 *   - No stable per-transaction id is exposed, which is why ingestion depends on
 *     the event fingerprint.
 */
export const ZKTECO_LEGACY_CONNECTOR: AttendanceConnectorDefinition = {
  connectorType: 'zkteco-legacy-tcp',
  provider: AttendanceProvider.ZKTECO,
  connectionMode: AttendanceConnectionMode.LOCAL_GATEWAY,
  label: 'ZKTeco Legacy Terminal',
  description:
    'Fingerprint and card terminals from the ZKTeco standalone range, connected over your local network through a DijiPeople gateway.',

  capabilities: [
    'READ_DEVICE_INFO',
    'READ_USERS',
    'READ_ATTENDANCE',
    'POLL_EVENTS',
    'DEVICE_TIME',
    'PUNCH_STATE',
    'WORK_CODE',
    'LOCAL_GATEWAY_REQUIRED',
    // Identity write-back. Declared, but unproven — see below.
    'WRITE_USERS',
  ],

  experimentalCapabilities: [
    {
      capability: 'WRITE_USERS',
      reason:
        'The SDK exposes a user write path, but it has not been executed against a physical terminal. Phase 2 validates it. Until then automatic provisioning must not use this connector unattended.',
    },
  ],

  configurationSchema: {
    fields: [
      {
        key: 'host',
        label: 'Device IP address',
        type: 'string',
        required: true,
        placeholder: '192.168.1.50',
        helpText: 'The address the gateway uses to reach this terminal.',
      },
      {
        key: 'port',
        label: 'Port',
        type: 'number',
        required: true,
        defaultValue: 4370,
        min: 1,
        max: 65535,
        helpText: 'ZKTeco terminals use 4370 unless it has been changed.',
      },
      {
        key: 'machineNumber',
        label: 'Device ID',
        type: 'number',
        required: true,
        defaultValue: 1,
        min: 0,
        max: 255,
        helpText: 'Shown on the terminal under Menu > Comm > Device ID.',
      },
      {
        key: 'commKey',
        label: 'Communication key',
        type: 'secret',
        required: false,
        secret: true,
        defaultValue: 0,
        helpText:
          'Set on the terminal under Menu > Comm > Security. Leave blank if no key is configured. Stored encrypted and never shown again.',
      },
      {
        key: 'expectedSerialNumber',
        label: 'Expected serial number',
        type: 'string',
        required: false,
        helpText:
          'Optional safety check. If set, a sync warns when the terminal that answers reports a different serial.',
      },
      {
        key: 'timezone',
        label: 'Device timezone',
        type: 'timezone',
        required: false,
        helpText:
          'The terminal reports wall-clock times with no timezone. Set this so punches can be resolved to a point in time.',
      },
      /**
       * How much of the terminal's stored history to bring into DijiPeople.
       *
       * This terminal family has no time-bounded read, so the gateway always
       * enumerates the whole log. What this setting controls is which of those
       * records are ADMITTED — the rest are fingerprinted locally and dropped,
       * which is what stops a first sync from importing four years of punches
       * into live attendance. Nothing is ever deleted from the device.
       */
      {
        key: 'initialSyncMode',
        label: 'Import history from',
        type: 'select',
        required: false,
        defaultValue: 'CURRENT_DATE',
        options: [
          { value: 'CURRENT_DATE', label: 'Today onwards' },
          { value: 'LAST_N_DAYS', label: 'A number of recent days' },
          { value: 'FROM_DATE', label: 'A specific date onwards' },
          { value: 'ALL_HISTORY', label: 'Everything stored on the terminal' },
        ],
        helpText:
          'Terminals keep years of punches. "Today onwards" is the safe choice for going live; the older records stay on the device untouched.',
      },
      {
        key: 'initialSyncDays',
        label: 'Days of history to import',
        type: 'number',
        required: false,
        defaultValue: 7,
        min: 1,
        max: 3650,
        helpText: 'Used when importing a number of recent days.',
      },
      {
        key: 'initialSyncFromDate',
        label: 'Import punches from (date)',
        type: 'string',
        required: false,
        placeholder: '2026-01-01',
        pattern: '\\d{4}-\\d{2}-\\d{2}',
        patternMessage: 'Enter a date as YYYY-MM-DD, for example 2026-01-01.',
        helpText:
          'Used when importing from a specific date. Interpreted in the terminal’s own local time, because that is the only clock its punches carry.',
      },
    ],
  },

  /**
   * A poll re-reads the device's whole history, so a short interval multiplies
   * load without yielding more information. The 15-minute floor is enforced,
   * not advisory.
   */
  recommendedSyncPolicy: {
    mode: 'POLL',
    recommendedIntervalValue: 30,
    recommendedIntervalUnit: AttendanceSyncIntervalUnit.MINUTES,
    minimumIntervalMinutes: 15,
    rationale:
      'This terminal family has no incremental read, so every sync downloads the full attendance history. Frequent polling adds load without adding data.',
  },

  supportsMultipleDevices: true,
  requiresGateway: true,

  notes: [
    'Requires a DijiPeople gateway on a Windows machine that can reach the terminal.',
    'Each sync reads the terminal’s full history; DijiPeople stores only punches it has not seen before.',
    'Choose how far back to import before going live. The terminal keeps its own records either way — DijiPeople never deletes or edits them.',
    'Biometric templates are never read, transferred or stored. Only employee identifiers and punch records are exchanged.',
    'Device user passwords returned by the terminal are discarded and never stored.',
  ],
};
